### PR TITLE
Added lazy events

### PR DESCRIPTION
**Pull request by DamianX**
I wanted to do a thing but then I had to use events and it reminded me of how awful they are, so I made this and I think it's pretty cool.
It's basically TG's signals but ~~shittier~~ simpler, and without the components.
They work like current `/event`s but:
- Any datum can hold any event.
- The list that holds the event handlers is a lazy list, as opposed to the old event system which requires creating a ton of `/event` datums most of which never get used.
- Event handlers are always called asynchronously, so sleeps in jukebox code won't cause the shuttle to split in two. Hypothetically speaking.
- It's documented and unit tested.

### CODE REVIEW

**Comment by DamianX** [comment]
I converted on_z_transition and on_post_z_transition as a proof of concept.
CC @ShiftyRail because you worked with those the most, I think

File: code/_hooks/events.dm [hunk 1]
```diff
@@ -51,3 +51,133 @@
...
+	var/list/event_handlers = registered_events[event_type]
+	if(!length(event_handlers))
+		// This datum does not have any handler registered for this event_type.
+		return
+	. = NONE
+	for(var/key in event_handlers)
+		var/list/handler = event_handlers[key]
+		var/objRef = handler[EVENT_HANDLER_OBJREF_INDEX]
+		var/procName = handler[EVENT_HANDLER_PROCNAME_INDEX]
+		. |= CallAsync(objRef, procName, arguments)
```
Comment (ShiftyRail) [inline code/_hooks/events.dm:98]: Why is it `|=` and not `+=` ?

File: code/_hooks/events.dm [hunk 2]
```diff
@@ -51,3 +51,133 @@
...
+	registered_events[event_type] -= key
+	if(!registered_events[event_type].len)
+		registered_events -= event_type
+	if(!registered_events.len)
+		registered_events = null
+
+#undef EVENT_HANDLER_OBJREF_INDEX
+#undef EVENT_HANDLER_PROCNAME_INDEX
+
+#if UNIT_TESTS_ENABLED
```
Comment (ShiftyRail) [inline code/_hooks/events.dm]: Why not put that in the unit testing folder ?

**Review by ShiftyRail — APPROVED** [review]
Looks alright to me, syntax a tad convoluted but that's BYOND

File: code/_hooks/events.dm (hunk 1 above)
Comment (DamianX) [reply code/_hooks/events.dm:98]: Why not? `|` lets you return bitflags, `+` would be less useful, I think. If the need for a different way of handling return values arises we'll change it but at the moment this is unused.

File: code/_hooks/events.dm (hunk 2 above)
Comment (DamianX) [reply code/_hooks/events.dm]: I kinda like having the tests right next to the code but whatever, I moved them

File: code/_hooks/events.dm (hunk 1 above)
Comment (ShiftyRail) [reply code/_hooks/events.dm:98]: Fair enough. Not like it would matter a lot in that context.

**Pull request merged**